### DISCUSSION TITLE
recognize not recognising glibc

### DISCUSSION THREAD

**Discussion by 31337-4554551n** [nextcloud/recognize#1163] [Q&A] [answered]
I am using unraid and docker, with linuxserver.io container.
It doesn't have glibc. So I went through a few steps to get it all installed, and by all accounts working. however recognize still won't see it there:

I created a special startup script to make sure glibc is installed
```
cat nextcloud-custom/custom-cont-init.d-fix/add-recognize-glibc-support 
#!/bin/sh
apk list | grep glibc
status=$?

if [ $status -gt 0 ]
then
wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://alpine-pkgs.sgerrand.com/sgerrand.rsa.pub
apk add /data/glibc-2.35-r1.apk
apk add /data/glibc-bin-2.35-r1.apk
fi
```
Not quite sure what the wget of the keys does, but I followed the instructions here:
https://github.com/sgerrand/alpine-pkg-glibc

Then I modified the PATH variable in the docker to prepend /usr/glibc-compat/bin: to the PATH that comes pre set with their container.

The installation works:
```
[custom-init] add-recognize-glibc-support: executing...
WARNING: opening from cache http://dl-cdn.alpinelinux.org/alpine/v3.20/main: No such file or directory
WARNING: opening from cache http://dl-cdn.alpinelinux.org/alpine/v3.20/community: No such file or directory
fetch http://dl-cdn.alpinelinux.org/alpine/v3.20/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.20/community/x86_64/APKINDEX.tar.gz
(1/1) Installing glibc (2.35-r1)
OK: 328 MiB in 318 packages
(1/4) Installing musl-obstack (1.2.3-r2)
(2/4) Installing libucontext (1.2-r3)
(3/4) Installing gcompat (1.1.0-r4)
ERROR: gcompat-1.1.0-r4: trying to overwrite lib/ld-linux-x86-64.so.2 owned by glibc-2.35-r1.
(4/4) Installing glibc-bin (2.35-r1)
Executing glibc-bin-2.35-r1.trigger
1 error; 331 MiB in 322 packages
[custom-init] add-recognize-glibc-support: exited 1
[ls.io-init] done.
nginx: [warn] the "listen ... http2" directive is deprecated, use the "http2" directive instead in /config/nginx/site-confs/default.conf:13
nginx: [warn] the "listen ... http2" directive is deprecated, use the "http2" directive instead in /config/nginx/site-confs/default.conf:14
```

Now in the container I am able to run
```
root@374a64c3838e:/# echo $PATH
/usr/glibc-compat/bin:/lsiopy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
root@374a64c3838e:/# ldd --version
ldd (GNU libc) 2.35
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
root@374a64c3838e:/# ldd --version
ldd (GNU libc) 2.35
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
```
However recognize still gives me the 
 Tensorflow WASM mode

WASM mode was activated automatically, because your machine does not support native TensorFlow operation:

    Your server uses musl libc

**Comment by marcelklehr** [comment]
Sometimes the automatic detection is wrong. If you're certain it should work, you can just disable WASM mode manually

**Reply by 31337-4554551n** [reply]
> Sometimes the automatic detection is wrong. If you're certain it should work, you can just disable WASM mode manually

I tried that, but then it complains that 

Could not load libtensorflow in Node.js. You can try to manually install libtensorflow or run in WASM mode. 


Is there some way to confirm it's using glibc?

**Reply by 31337-4554551n** [reply]
What is Node.js there for? Ditto tenserflow, is it neccesary? Do I also need to find a way to install tenserflow on alpine?

**Reply by marcelklehr** [reply, marked as answer]
Many people have tried making it work with glibc on alpine linux, without success, afaik.
Node.js is there to run tensorflow.js which uses tensorflow. They're both necessary.
You can try building tensorflow yourself on alpine linux, I guess. The fact that this is so hard is the reason that recognize will automatically switch to WASM mode when you install it on alpine.

**Reply by 31337-4554551n** [reply]
Yea, i managed to get glibc working as there's a repo for it for alpine, but then you get to the tenserflow errors :(
